include/drivers/video.h

```diff
@@ -0,0 +1,317 @@
+/**
+ * @file
+ *
+ * @brief Public APIs for Video.
+ */
+
+/*
+ * Copyright (c) 2019 Linaro Limited.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_VIDEO_H_
+#define ZEPHYR_INCLUDE_VIDEO_H_
+
+#include <device.h>
+#include <stddef.h>
+#include <zephyr.h>
+
+#include <zephyr/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+enum video_device_type {
+	video_CAPTURE_DEVICE,
+	video_OUTPUT_DEVICE,
+	video_CONTROL_DEVICE,
+};
+
+struct video_device_config {
+	enum video_device_type type;
+};
+
+struct video_format {
+	u32_t pixelformat;
+	u32_t width;
+	u32_t height;
```

Comment:
> [ Camera ] => [ Software scaler/filter/color space converter ] => [ HW Encoder ] => Storage
> 
 
> Questions:
> 
>     * Each of the device provides its own video frame allocator. Am I supposed to reallocate frame and copy its contents manually between each processing step or I can pass the frame directly?

You only need to allocate buffer one time, and you can move them without copy/reallocation.
For now, allocate() support is optional and would be mainly implemented by  platform interface video drivers (parallel bus/CSI), because they know the low-level requirements (e.g. memory area, alignment, etc...) But we could also imagine creating a generic video allocation function (e.g. based on k_mem_pool and) which would take care of allocating buffer according to the all pipelined video device requirements (camera, hw encoder, dispaly, etc...).

>     * I assume that I enqueue uncompressed frame to the HW encoder and dequeue the compressed one. Since most of the video compression algorithms do not support compression in place, the HW encoder will allocate the output frame. Who is responsible for freeing the input frame? If user of the API, how he/she could know when the frame is no longer used. If the HW encoder driver, how it could know which driver is owner of the frame and which version of video_api_release_frame_t should be called?

You're right, Several possibilities here, but I would add a video_buffer type (e.g. CAPTURE & OUTPUT), in case of hw encoder which is a mem 2 mem device the user needs to 'queue' CAPTURE and OUTPUT buffers, the encoder converts from CAPTURE to OUTPUT buffers, on 'dequeue' CAPTURE buffers are then re-used for capture (e.g. camera, read from file...), OUTPUT buffer are the converted ones and can be e.g. sent over the network, written to SDCARD... Anyway, an enqueued buffer will always be dequeued at some point.

>     * Some of the video encoders need some kind of out of band data describing properties of the compressed stream (the H264 is good example). How we are supposed to retrieve this data from the encoder and pass it to the decoder? Will we have concept of the stream in the API?

Well, I try to start with something simple for now, but really open to improvement, adding a kind of stream context or/and some meta-data makes sense.

Reply:
> For now, allocate() support is optional and would be mainly implemented by platform interface video drivers (parallel bus/CSI), because they know the low-level requirements (e.g. memory area, alignment, etc...)

IMHO there should be an information about way how the video buffer is allocated inside video buffer structure. This way, any component will be able to free the buffer without knowing origin:
```
video_buffer_free(*buffer)
{
    buffer->free(buffer);
}
```

> I would add a video_buffer type like CAPTURE & OUTPUT. In case of camera user enqueue CAPTURE buffers which become OUTPUT on dequeue

I don't think that such type should exist. Buffer is just a buffer. Only HW has some limitations which must be taken under account during allocation. Ideally one allocated buffer should be used by all components in the processing pipeline. If you allow per-device buffers, you will eventually end in memory copy.

For example: The input might need buffer aligned to 4 bytes, but output device requires 16 byte alignment. If you allocate buffer from input device, you will have to copy the data. But if you allocate it from output, this will be not needed (assuming that the alignment is only problem here).

In general, I would suggest looking into OpenMAX IL and Gstreamer specifications. Both are too heavy for Zephyr, but it some problems are already solved there (like zero-copy buffer management, stream context etc.).

Reply:
Well, I modified the API and moved video_buffer allocation from device driver to a generic video allocation handler. This simplifies the API, up to the platform/board config to overwrite default buffer alignment if requested.

Reply:
@pizi-nordic, So I finally had some time to review different video frameworks and changed my mind on this, we can indeed see the video buffer as the transport mechanism only, without any requirement on content. The format itself is defined by the endpoint/port.

Reply:
I agree to this change, as a good design choice, memory should be handled in the high level system rather than the low level function.